src/worker.go

```diff
@@ -69,71 +85,99 @@ func (w *Worker) Start() {
 }
 
 // Work on a single job
-func (w *Worker) Work(job *Job) error {
-	// We have received a work request.
-	if verbose {
-		log.Printf("(%d worker) Received a job request!\n", w.ID)
-	}
+func (w *Worker) Work(action *EventAction) error {
 	if !storageClient.IsBufferedStorage() {
-		// Convert the message to string
-		msg, err := storageClient.GetConverter()(job.Event)
-		if err != nil {
-			rerr := fmt.Errorf("(%d worker) Encoding message is failed (%d attempt): %s", w.ID, job.Attempt, err.Error())
-			job.MarkAsFailed(w.RetryAttempt)
-			return rerr
+		// Save messages
+		if err := w.Save(action); err != nil {
+			return err
 		}
 
-		// Save message immediately.
-		if err := storageClient.Save(msg); err != nil {
-			rerr := fmt.Errorf("(%d worker) Saving message is failed (%d attempt): %s", w.ID, job.Attempt, err.Error())
-			job.MarkAsFailed(w.RetryAttempt)
-			return rerr
-		}
 	} else {
 		// Add message to the buffer if the storge is a buffered writer
-		w.AddEventToBuffer(job.Event)
+		w.AddEventToBuffer(action.GetEvent())
 
 		// Continue if the buffer is not full
 		if !w.IsBufferFull() {
 			return nil
 		}
 
-		// Convert messages to string
-		msg, err := storageClient.GetBatchConverter()(w.BufferedEvents)
-		if err != nil {
-			w.IncreasePenalty()
-			return fmt.Errorf("(%d worker) Batch converting buffered messages is failed with %d records: %s", w.ID, len(w.BufferedEvents), err.Error())
+		if verbose {
+			log.Printf("(%d worker) Saving buffered messages started", w.ID)
 		}
+
 		// Save messages
-		if err := storageClient.Save(msg); err != nil {
-			w.IncreasePenalty()
-			return fmt.Errorf("(%d worker) Saving buffered messages is failed with %d records: %s", w.ID, len(w.BufferedEvents), err.Error())
+		if err := w.SaveBatch(); err != nil {
+			return err
+		}
+
+		if verbose {
+			log.Printf("(%d worker) Saving buffered messages was finished", w.ID)
 		}
-		w.ResetBuffer()
 	}
 	return nil
 }
 
+// Save Buffered messages
+func (w *Worker) SaveBatch() error {
+	// Convert messages to stringdefer
```

Comment:
Please do not make a save when the buffer is empty but please update the last saving date.